apps/explorer/lib/explorer/chain/smart_contract/proxy.ex

```diff
@@ -480,17 +480,56 @@ defmodule Explorer.Chain.SmartContract.Proxy do
   @doc """
   Returns combined ABI from proxy and implementation smart-contracts
   """
-  @spec combine_proxy_implementation_abi(any(), any()) :: SmartContract.abi()
-  def combine_proxy_implementation_abi(smart_contract, options \\ [])
+  @spec combine_proxy_implementation_abi(any(), list(), boolean(), any()) :: SmartContract.abi()
+  def combine_proxy_implementation_abi(
+        smart_contract,
+        proxy_implementation_addresses_tuple_list \\ [],
+        fetch_proxy?,
+        options \\ []
+      )
 
-  def combine_proxy_implementation_abi(%SmartContract{abi: abi} = smart_contract, options) when not is_nil(abi) do
-    implementation_abi = Proxy.get_implementation_abi_from_proxy(smart_contract, options)
+  def combine_proxy_implementation_abi(
+        %SmartContract{abi: abi} = smart_contract,
+        proxy_implementation_addresses_tuple_list,
+        fetch_proxy?,
+        options
+      )
+      when not is_nil(abi) do
+    implementation_abi =
+      get_implementation_abi(smart_contract, options, proxy_implementation_addresses_tuple_list, fetch_proxy?)
 
     if Enum.empty?(implementation_abi), do: abi, else: implementation_abi ++ abi
   end
 
-  def combine_proxy_implementation_abi(_, _) do
-    []
+  def combine_proxy_implementation_abi(smart_contract, proxy_implementation_addresses_tuple_list, fetch_proxy?, options) do
+    get_implementation_abi(smart_contract, options, proxy_implementation_addresses_tuple_list, fetch_proxy?)
+  end
+
+  defp get_implementation_abi(smart_contract, options, proxy_implementation_addresses_tuple_list, fetch_proxy?) do
+    if fetch_proxy? do
+      Proxy.get_implementation_abi_from_proxy(smart_contract, options)
+    else
+      proxy_implementation_addresses_tuple =
+        proxy_implementation_addresses_tuple_list
+        |> Enum.find(fn {proxy_address_hash, _implementations} ->
+          proxy_address_hash.bytes == (smart_contract.address_hash && smart_contract.address_hash.bytes)
```

Comment:
Maybe it's better to compare them as structs, to avoid dealing with nil values and destructuring
```suggestion
          proxy_address_hash == smart_contract.address_hash
```